keps/sig-storage/3751-volume-provisioned-io/README.md

```diff
@@ -0,0 +1,1338 @@
+
+Before:
+
+```
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: test-pv-claim
+spec:
+  storageClassName: csi-sc-example
+  VolumeAttributesClassName: silver
+  accessModes:
+    - ReadWriteOnce
+  resources:
+    requests:
+      storage: 64Gi
+```
+
+After:
+
+```
+apiVersion: storage.k8s.io/v1alpha1
+kind: VolumeAttributesClass
+metadata:
+  name: gold
+parameters:
+  iops: "1000"
+  throughput: "100MiB/s"
+```
+
+```
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: test-pv-claim
+spec:
+  storageClassName: csi-sc-example
+  VolumeAttributesClassName: gold
+  accessModes:
+    - ReadWriteOnce
+  resources:
+    requests:
+      storage: 64Gi
+```
+
+#### Administrator Quota Restrictions
+
+A cluster admin wants to control costs while giving application developers the freedom to optimize their applications. They set a per VolumeAttributesClass limit to the maximum count of PVCs that can be specified in a cluster ResourceQuota. When an application dev modifies a PVC to request a higher tier of VolumeAttributesClass but there is no quota, the request is rejected. As the ResourceQuota is a cluster-scoped object, only the cluster admin and not application devs can change limits.
+
+
+### Notes/Constraints/Caveats (Optional)
+
+The parameters in VolumeAttributesClass are opaque. This gives different cloud providers flexibility but at the same time it is up to the specific driver to verify the values set in the parameters. The suggestion is to set immutable parameters in StorageClass and mutable parameters in VolumeAttributesClass.
+
+### Risks and Mitigations
+
+Users may configure both parameters in StorageClass and parameters in VolumeAttributesClass differently. The mitigation here is to provide a guideline of solving conflicts. If the MODIFY\_VOLUME capability is present, VolumeAttributesClass should be honored.
+
+## Design Details
+
+As part of this proposal, we are proposing:
+
+
+### 1. Add Create/Delete VolumeAttributesClass support in Kubernetes including vac\_admission\_controller and vac\_protection\_controller for deletion protection. 
+
+
+#### VolumeAttributesClass Deletion Protection
+
+While a VolumeAttributesClass is referenced by any PVC, we will prevent the object from being deleted by adding a finalizer([reference](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/storage/storageobjectinuseprotection/admission.go)). 
+
+The VACObjectInUseProtection admission controller sets the finalizer on all VolumeAttributesClasses. VACProtectionController removes the finalizer when it's not referenced. This prevents users from deleting a VolumeAttributesClass that's used by a PVC.
+
+There are a few conditions that will trigger add/remove pvc finalizers in the VolumeAttributesClass:
+
+
+
+1. PVC created with a VolumeAttributesClass
```

Comment:
The PVC should be in the state of pending.

Reply:
So the PVC will be created, but then the admission controller would have already run and won't run again. So in this scenario, the VAC protection controller would have to handle adding the finalizer. But then I think it will be prone to races. 

@jsafrane I know we initially put the pvc/pv protection controller as a separate controller to not make the pv controller more complicated, but I wonder if in this case it would be simpler to have the finalizer logic inline with pv controller and external-resizer, which are the ones that gate the CSI calls. I think import PV case may still be tricky though to be race free.